constraint/pkg/client/matchers.go

```diff
@@ -0,0 +1,173 @@
+
+	c.matchers[key.target] = target
+}
+
+// Remove deletes the Matcher for the Constraint with kind and name.
+// Returns normally if no entry for the Constraint existed.
+func (c *constraintMatchers) Remove(key matcherKey) {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	if len(c.matchers) == 0 {
+		return
+	}
+
+	target := c.matchers[key.target]
+	if len(target.matchers) == 0 {
+		return
+	}
+
+	target.Remove(key)
+
+	if len(target.matchers) == 0 {
+		delete(c.matchers, key.target)
+	} else {
+		c.matchers[key.target] = target
+	}
+}
+
+// RemoveKind removes all Matchers for Constraints with kind.
+// Returns normally if no entry for the kind exists for any target.
+func (c *constraintMatchers) RemoveKind(kind string) {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	if len(c.matchers) == 0 {
+		return
+	}
+
+	for name, target := range c.matchers {
+		target.RemoveKind(kind)
+
+		if len(target.matchers) == 0 {
+			// It is safe to delete keys from a map while traversing it.
+			delete(c.matchers, name)
+		} else {
+			c.matchers[name] = target
+		}
+	}
+
+	delete(c.matchers, kind)
+}
+
+// ConstraintsFor returns the set of Constraints which should run against review
+// according to their Matchers. Returns a map from Kind to the names of the
+// Constraints of that Kind which should be run against review.
+//
+// Returns errors for each Constraint which was unable to properly run match
+// criteria.
+func (c *constraintMatchers) ConstraintsFor(targetName string, review interface{}) (map[string][]string, error) {
+	c.mtx.RLock()
+	defer c.mtx.RUnlock()
+
+	result := make(map[string][]string)
+	target := c.matchers[targetName]
+	errs := errors.ErrorMap{}
+
+	for kind, kindMatchers := range target.matchers {
+		var resultKindMatchers []string
+
+		for name, matcher := range kindMatchers {
+			if matches, err := matcher.Match(review); err != nil {
+				// key uniquely identifies the Constraint whose matcher was unable to
+				// run, for use in debugging.
+				key := fmt.Sprintf("%s %s %s", targetName, kind, name)
+				errs[key] = err
+			} else if matches {
+				resultKindMatchers = append(resultKindMatchers, name)
+			}
+		}
+
+		sort.Strings(resultKindMatchers)
+		result[kind] = resultKindMatchers
+	}
+
+	if len(errs) > 0 {
+		return nil, &errs
+	}
+
+	return result, nil
+}
+
+// targetMatchers are the Matchers for Constraints for a specific target.
+// Not threadsafe.
+type targetMatchers struct {
+	matchers map[string]map[string]constraints.Matcher
+}
+
+func (t *targetMatchers) Add(key matcherKey, matcher constraints.Matcher) {
+	if t.matchers == nil {
+		t.matchers = make(map[string]map[string]constraints.Matcher)
+	}
+
+	kindMatchers := t.matchers[key.kind]
+	if kindMatchers == nil {
+		kindMatchers = make(map[string]constraints.Matcher)
+	}
+
+	kindMatchers[key.name] = matcher
+	t.matchers[key.kind] = kindMatchers
+}
+
+func (t *targetMatchers) Remove(key matcherKey) {
+	kindMatchers := t.matchers[key.kind]
+	if len(kindMatchers) == 0 {
```

Comment:
Should we do

```
kindMatchers, ok := t.matchers[key.kind]
```

and use `ok` here to check for existence?

That way we have GC regardless of whether the map happens to be empty?

Reply:
Done.